Classifier: Consolidate store subject observers

eatyourgreens opened the pull request:
Consolidate the subject observers from the feedback, workflow steps and classifications stores into one function in the root store. When the subject changes, reset the workflow steps, create a new classification, then reset subject feedback. I think this will fix #1521 by making sure the first task is selected before the classification is created.

This PR also fixes a couple of infinite loops that showed up in the Tutorial Store tests (infinitely requesting new subjects) and the UPP Store tests (infinitely requesting a new workflow.) 

I've also moved the Feedback `onAction` listener up to the root in order to mute warnings from that store. Classification and Mark `onAction` listeners are replaced with `beforeDestroy` actions to remove those warnings too. Otherwise, I was finding the console so noisy that failing tests were impossible to debug.

Package:
lib-classifier

Closes #1521.

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?

## Components
- [ ] Has a storybook story been created or updated?
- [ ] Is the component accessible? 
  - [ ] Can it be used with a screen reader? [BBC guide to testing with VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - [ ] Can it be used from the keyboard? [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - [ ] Is it passing accessibility checks in the storybook?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `yarn panic && yarn bootstrap` or `docker-compose up --build` and app works as expected?

## Publishing

- [ ] Is the changelog updated?
- [ ] Are the dependencies updated for apps and libraries that are using the newly published library?

## Post-merging

- [ ] Did the app deploy to https://frontend.preview.zooniverse.org/projects/:project-name/:owner or https://frontend.preview.zooniverse.org/about?
- [ ] Is the new feature working or bug now fixed?
  - [ ] Is there a Talk or blog post written to announce the new feature(s)?
- [ ] Is the design working across browsers (Firefox, Chrome, Edge, Safari) and mobile?
  - [ ] Is this approved by our designer?
- [ ] Is this ready for production deployment?

eatyourgreens commented:
Testing this is throwing up a bunch of infinite loops in the mock Root Stores used in the UPP tests. I'm looking into that to see if I can figure out what's going on.

eatyourgreens commented:
UPP is in a loop because it constantly requests a workflow but receives nothing.

eatyourgreens commented:
Tests are fixed now so I think this is ready for review.

I've also moved the feedback onAction listener up to the root in order to mute some of the onAction warnings.

shaunanoordin commented:
## Quick Review

Functionally, this PR looks legit - the changes to the code correctly ensures that annotations are made in the correct order, _for the 2nd Classification onwards._ (The bug does not affect the 1st Classification submitted. See #1521 for deets.)

e.g. for a Workflow with 3 Tasks, the annotation values submitted to Panoptes _on this branch_ are in the form of `[ { task: 'T0', ... }, { task: 'T1', ... }, { task: 'T2', ... } ]`, which is expected.

This is instead of the form of `[ { task: 'T2', ... }, { task: 'T0', ... }, { task: 'T1', ... } ]`, which is what's going on in `master` (6a913857ace00b1f1abb83678713d66397395dac)

This PR gets a 👍 from a functional point of view, though I'm going to take my time to grok the actual code changes.

eatyourgreens commented:
I'm going to try changing this to run `onSubjectReady` too. I think that would make it more clear that we're resetting the workflow etc. after a new subject loads.

eatyourgreens commented:
I've updated this to reset the workflow, create a new classification and reset feedback on subject ready. I've also added tests that check the store state after `onSubjectReady` runs.

Mocking the root store for the tests was a lot harder than I expected. There was a lot of reading through store code to figure what needed to be set and which actions should be called to set up the initial store state. It turns out that the root store can't be created directly from a snapshot. You must use `ResourceStore.setResource` and `ResourceStore.setActive` to set up workflows and subjects. That could be improved, maybe as part of #1554.

eatyourgreens commented:
Is there a project where we can test this with feedback enabled?

eatyourgreens commented:
The classification store tests for 'with valid feedback > classification metadata' are throwing MST warnings.
```
Error: [mobx-state-tree] You are trying to read or write to an object that is no
longer part of a state tree. (Object type: 'ClassificationMetadata',
Path upon death: '/classifications/resources/ck7lzu84s0066lktw7beca3gn/metadata',
Subpath: 'feedback', Action: '').
Either detach nodes first, or don't use objects after removing / replacing
them in the tree.
```

I'm not seeing any warnings in the browser so I'm going to put that down to something being not quite right in the mocks for those tests. I can't immediately see what it is.

eatyourgreens commented on packages/lib-classifier/src/store/ClassificationStore.spec.js:
```diff
@@ -245,7 +255,7 @@ describe('Model > ClassificationStore', function () {
       // Why is this test here?
       // The observer is in the feedback store
```
I think I’ve moved this observer in this PR.

srallen replied:
Non-blocking comment: Yep. Will these specs move to the `RootStore` spec file? At the very least the comment about the observer can be removed.

eatyourgreens replied:
This should move up to the root store tests. So much is being stubbed out here that I find it hard to follow what's really being tested. When these tests are moved, I'd suggest rewriting them to remove the stubs and test expected feedback state after a classification completes instead.

srallen commented on packages/lib-classifier/src/store/ClassificationStore.spec.js:
```diff
@@ -111,8 +117,10 @@ describe('Model > ClassificationStore', function () {
       })
-      rootStore.subjects.setResource(subjectsStub[0])
-      rootStore.subjects.setActive(subjectsStub[0].id)
+      rootStore.subjectViewer.onSubjectReady()
```
Same comment about calling `onSubjectReady` here.

srallen commented on packages/lib-classifier/src/store/RootStore.js:
```diff
@@ -33,6 +34,49 @@ const RootStore = types
+      const subjectDisposer = autorun(() => {
+        onAction(self, (call) => {
+          if (call.name === 'onSubjectReady') {
+            onSubjectReady()
```
Nice. This is easier to read and clearer on what is triggering the reset.

eatyourgreens commented on packages/lib-classifier/src/store/ClassificationStore.spec.js:
```diff
@@ -66,8 +70,10 @@ describe('Model > ClassificationStore', function () {
       })
-      rootStore.subjects.setResource(subjectsStub[0])
-      rootStore.subjects.setActive(subjectsStub[0].id)
+      rootStore.subjectViewer.onSubjectReady()
```
This broke the tests when I ran it inside `setupStores`, after the active subject is set there, so I moved it out. I'll take another look.

eatyourgreens replied:
I remember. This is here to work around a problem with the feedback tests. I've rewritten the setup function, and added a comment about the workaround.

eatyourgreens commented:
I think the drawing task crash might be happening when the subject renders `onReady` but before the new classification has been created. So it's trying to render annotations from the submitted classification.

eatyourgreens commented:
>I think this can be fixed by checking for the SubjectViewerStore's loading state in InteractionLayerContainer to be success before rendering.

That worked, thanks. I changed `enableInteractionLayer` to check the loading state too.

shaunanoordin commented on packages/lib-classifier/src/components/Classifier/components/TaskArea/components/Tasks/Tasks.js:
```diff
@@ -41,21 +39,21 @@ class Tasks extends React.Component {
+    const { classification, isThereTaskHelp, subjectReadyState, step, theme } = this.props
     const ready = subjectReadyState === asyncStates.success
-    if (classification && tasks.length > 0) {
+    if (classification && step.tasks.length > 0) {
```
[minor] there's no instance where `step` can be undefined, right?

eatyourgreens replied:
I don't think so. I _think_ the model always has a `tasks` property, which defaults to `[]`.

shaunanoordin commented on packages/lib-classifier/src/store/RootStore.js:
```diff
@@ -33,6 +34,49 @@ const RootStore = types
+    // Private methods
+    function onSubjectReady () {
+      const { classifications, feedback, projects, subjects, workflows, workflowSteps } = self
+      const subject = tryReference(() => subjects?.active)
```
[aside] Wait, we have `?.` existence checkers?? This is cool, I didn't realise we had this - where's it coming from? Babel? ES2020?

srallen replied:
[Optional chaining operator](https://github.com/tc39/proposal-optional-chaining)! It's stage 4 now, so it's part of the default babel support and should be coming to the browser soon!

eatyourgreens commented:
>(If I'm understanding the cause of issue #1521 is that there's some sort of race condition where the indvidualistic sub-stores are basically performing their cleanup operations regardless of what the other sub-stores are doing, so e.g. annotations stores are resetting before the task stores, or some sort of similar madness.)

I don't fully understand it, but I think this is the case. We don't have any control over the order in which the individual subject observers run in the workflow steps, classifications and feedback models. My idea was to consolidate them into a single function and run them in a prescribed order.
